### PR TITLE
Introducing Endorsement API

### PR DESCRIPTION
**Pull request by yogeshbdeshpande**

### CODE REVIEW

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
+openapi: '3.0.0'
+
+info:
+  title: attestation endorsement provisioning
+  description: For further details, see datatracker.ietf.org/doc/draft-birkholz-corim
+  version: '1.0.0alpha'
+servers:
+  - url: 'https://veraison.example/endorsement-provisioning/v1'
+paths:
+  /CoRIM:
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:10]: Consider changing to Submit - Path should be an Action.

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    post:
+      description: >
+        Submit a single CoRIM or a batch of CoRIM payloads using this API.
+      requestBody:
+        required: true
+        content:
+          application/corim:
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:17]: Refer IANA Consideration: Content Type:

application/corim+cbor

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          description: >
+            The submission has been successful and will be processed
+            asynchronously.
+          content:
+            application/json:
+              schema:
+                $ref: './schemas/components.yaml#/components/schemas/CoRIMResponse'
+        default:
+          $ref: './schemas/components.yaml#/components/responses/Error'
+  /CoRIM/{CoRIMID}:
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:39]: postpone GET for now and also DELETE

Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/endorsement-provisioning.yaml:39]: This is an Additional Feature.. Do we want to expose it?

File: docs/api/endorsement-provisioning/schemas/components.yaml [hunk 4]
```diff
@@ -0,0 +1,105 @@
+components:
+  schemas:
+    CoRIMData:
+      required:
+        - payload
+      properties:
+        payload:
+          type: string
+          format: byte
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/schemas/components.yaml:9]: Check the type : Most likely it is byte, remove the description.

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+              $ref: './schemas/components.yaml#/components/schemas/CoRIMData'  
+      responses:
+        '200':
+          description: >
+            A new CoRIM or a batch of CoRIM has been successfully provisioned inside 
+            a Verifier. The transaction is complete.
+          content:
+            application/json:
+              schema:
+                $ref: './schemas/components.yaml#/components/schemas/CoRIMStatus'
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:28]: Remove this not needed

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        '202':
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:29]: Check if it is 201 ?

File: docs/api/endorsement-provisioning/schemas/components.yaml [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+          description:
+            base64 encoded bytes containing single CoRIM or a CoRIM batch
+          example: xyzpqrhfassddfrxcvbngdstygz=
+    CoRIMResponse:
+      required:
+        - status
+        - handle
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/schemas/components.yaml:16]: no need to have handle here..

File: docs/api/endorsement-provisioning/schemas/components.yaml [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        - expiry
+      properties:
+        status:
+          type: string
+          enum:
+            - processing
+            - complete
+            - failed
+          description: >
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/schemas/components.yaml:25]: Add details for failed status.

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            The submission is successful and will be processed
+            asynchronously. A new session resource is created and its URL is returned in the
+            Location header.  The resource will be kept alive for the specified
+            time-to-live.  The new resource comprises the handle to be used to check the status
+            of the ongoing request.
+          headers:
+            Location:
+              schema:
+                type: string
+                format: uri
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:40]: Full uri will be retuerbned in the link header!

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+                URL of the newly created session resource to be used in
+                subsequent status check.  (The current content of the
+                resource is returned in the response body.)
+          content:
+            application/json:
+              schema:
+                $ref: './schemas/components.yaml#/components/schemas/CoRIMResponse'
+        default:
+          $ref: './schemas/components.yaml#/components/responses/Error'
+  /session/{handle}:
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:51]: /submissions/{sessionId}

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+                type: string
+                format: uri
+              description:
+                URL of the newly created session resource to be used in
+                subsequent status check.  (The current content of the
+                resource is returned in the response body.)
+          content:
+            application/json:
+              schema:
+                $ref: './schemas/components.yaml#/components/schemas/CoRIMResponse'
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:48]: |Rename CoRIMResponse to remove CoRIM Response.

Just make it Response

Make it abstract which is Like request, submission and submission response

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+      parameters:
+        - $ref: '#/components/parameters/handle'
+      responses:
+        '200':
+          description: >
+            The handle or identifier which was returned at the time of CoRIM payload submission
+          content:
+            application/json:
+              schema:
+                $ref: './schemas/components.yaml#/components/schemas/CoRIMStatus'
```
Comment (yogeshbdeshpande) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:64]: SessionResource will be returned here: So we are now returnfing the resource, as returned in POST 201 response!

File: docs/api/endorsement-provisioning/README.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+[Concise Reference Integrity Manifest](https://datatracker.ietf.org/doc/draft-birkholz-rats-corim/).
+
+## Provisioning API
+
+Provisioning API provides a facility to provision either Reference Values or Endorsed Values, using a 
+single CoRIM or a batch of CoRIM(s) in a single payload.
+
+* A client who intends to submit a CoRIM or a batch of CoRIM invokes a `POST` on a submit path;
+* If the transaction is complete then a `(200)`response is returned to the client. This is an indication
+  that the submitted CoRIM or batch has been successfully provisioned in a verifier. The ongoing transaction is now complete.
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/README.md:15]:
```suggestion
The ongoing transaction is now complete.
```

File: docs/api/endorsement-provisioning/README.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+* For large submission with multiple CoRIMs, the provisioning processing may happen in background. In this case,
+  Client is returned a `(202)` response, with a status of `processing`. Also
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/README.md:17]: we said this was `201` with the Link header populated with the session resource?

File: docs/api/endorsement-provisioning/README.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+Provisioning API provides a facility to provision either Reference Values or Endorsed Values, using a 
+single CoRIM or a batch of CoRIM(s) in a single payload.
+
+* A client who intends to submit a CoRIM or a batch of CoRIM invokes a `POST` on a submit path;
+* If the transaction is complete then a `(200)`response is returned to the client. This is an indication
+  that the submitted CoRIM or batch has been successfully provisioned in a verifier. The ongoing transaction is now complete.
+* For large submission with multiple CoRIMs, the provisioning processing may happen in background. In this case,
+  Client is returned a `(202)` response, with a status of `processing`. Also 
+  client is returned with a session handle to subsequently poll the request for only a given amount of time.
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/README.md:18]: it's not a handle, it's a resource.

File: docs/api/endorsement-provisioning/README.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+```
+
+## Asynchronous submission
+
+```
+ o        (1) POST           .-------------.
+/|\ ------------------------>| /submit     |
+/ \ \ <----------------------|             |
+ \ \ \      201 ( Session )  '-------------'
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/README.md:40]: ok, yes, it was 201 :-)

File: docs/api/endorsement-provisioning/README.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+### Synchronous POST
+
+* Client submits the endorsement provisioning request
+* Server responds with response code `200` indicating successful submission. 
+  The transation is now complete
+```
+>> Request:
+  POST /endorsement-provisioning/v1/submit/?payload=56789abcdesg
+  Host: veraison.example
+  Accept: application/rim+cbor
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/README.md:59]: it's the server that accepts CoRIM, not the client.

This should be `Content-Type` instead.

(And we could have a MIME type for the session resource that the client could `Accept`.)

File: docs/api/endorsement-provisioning/README.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+<< Response:
+  HTTP/1.1 200 OK
+  Content-Type: application/json
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/README.md:63]: I am not sure what JSON is returned here, but if there is something in the response body then we should probably define a specific MIME type for it.

File: docs/api/endorsement-provisioning/README.md [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+```
+>> Request:
+  POST /endorsement-provisioning/v1/submit/?payload=489765abcdesgfdwdunisdufbiubwjndisjd
+  Host: veraison.example
+  Accept: application/rim+cbor
+
+<< Response:
+  HTTP/1.1 201 Created
+  Content-Type: application/session-response+json
+  Location: https://veraison.example/endorsement-provisioning/v1/session/1234567890
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/README.md:81]: Location can be a relative URL, it's not necessary to make it absolute.

File: docs/api/endorsement-provisioning/README.md [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+    "expiry": "2030-10-12T07:20:50.52Z"
+  }
+```
+
+### Polling the session resource
+```
+>> Request:
+  GET /endorsement-provisioning/v1/session/1234567890
+  Host: veraison.example
+  Accept: application/json
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/README.md:94]: we should `Accept` the session resource's MIME type here

File: docs/api/endorsement-provisioning/README.md (hunk 16 above)
Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/README.md:41]: Correct, my bad, i have corrected in the text!

File: docs/api/endorsement-provisioning/README.md [hunk 21]
```diff
@@ -0,0 +1,104 @@
+# Endorsement Provisioning Interface
+
+The APIs described here are meant to illustrate the provisioning of 
+Endorsements based on RATS architecture using the Data Model as
+illustrated in 
+[Concise Reference Integrity Manifest](https://datatracker.ietf.org/doc/draft-birkholz-rats-corim/).
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/README.md]: I have done the necessary rewording as I agree that we need to be format agnostic!

File: docs/api/endorsement-provisioning/README.md [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+## Provisioning API
+
+Provisioning API provides a facility to provision either Reference Values or Endorsed Values, using a 
+single CoRIM or a batch of CoRIM(s) in a single payload.
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/README.md]: Nice, taken the change and fine tuned it!

File: docs/api/endorsement-provisioning/README.md [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+```
+
+### Synchronous POST
+
+* Client submits the endorsement provisioning request
+* Server responds with response code `200` indicating successful submission. 
+  The transation is now complete
+```
+>> Request:
+  POST /endorsement-provisioning/v1/submit/?payload=56789abcdesg
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/README.md:58]: Yes, agree, corrected it now!

File: docs/api/endorsement-provisioning/README.md (hunk 18 above)
Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/README.md:64]: I do not see anything returned here ATM, so removed the json line here.

File: docs/api/endorsement-provisioning/README.md (hunk 19 above)
Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/README.md:82]: I agree, have corrected it now!

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 24]
```diff
@@ -0,0 +1,71 @@
+openapi: '3.0.0'
+
+info:
+  title: attestation endorsement provisioning
+  description: For further details, see datatracker.ietf.org/doc/draft-birkholz-corim
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:5]: not sure this applies:

```suggestion
  description: For further details, see datatracker.ietf.org/doc/draft-birkholz-corim
```

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+  /submit:
+    post:
+      description: >
+        Submit a single CoRIM or a batch of CoRIM payloads using this API.
+      requestBody:
+        required: true
+        content:
+          application/rim+cbor:
+            schema:
+              $ref: './schemas/components.yaml#/components/schemas/Submission'
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:19]: I am not sure we need a schema here: it's just binary data and a content-type

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+        required: true
+        content:
+          application/rim+cbor:
+            schema:
+              $ref: './schemas/components.yaml#/components/schemas/Submission'  
+      responses:
+        '200':
+          description: >
+            A new CoRIM or a batch of CoRIM has been successfully provisioned inside 
+            a Verifier. The transaction is complete.
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/endorsement-provisioning.yaml:24]:
```suggestion
            A new CoRIM or a batch of CoRIM has been successfully provisioned in 
            the Verifier. The transaction is complete.
```

File: docs/api/endorsement-provisioning/schemas/components.yaml [hunk 27]
```diff
@@ -0,0 +1,87 @@
+components:
+  schemas:
+    Submission:
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/schemas/components.yaml:3]: as noted above, this seem redundant

File: docs/api/endorsement-provisioning/schemas/components.yaml [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+    SessionResponse:
+      required:
+        - status
+        - expiry
+      properties:
+        status:
+          type: string
+          enum:
+            - processing
+            - completed
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/schemas/components.yaml:21]:
```suggestion
            - complete
```

File: docs/api/endorsement-provisioning/schemas/components.yaml [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+          format: date-time
+          description: >
+            time (UTC ISO 8601 timestamp) at which the resource associated with
+            the endorsement provisioning session expires and may no longer exist. The
+            client MUST assume that the endorsement provisioning session resource is
+            garbage-collected at the specified time.
+        details:
+          type: string
+          description: >
+            Optional details that explains the cause when provisioning status is `failed`
```
Comment (thomas-fossati) [inline docs/api/endorsement-provisioning/schemas/components.yaml:38]:
```suggestion
            Optional field to provide more details about the failure cause when provisioning status is `failed`
```

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml (hunk 25 above)
Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/endorsement-provisioning.yaml:19]: No it is good to keep it like that for consistency!

Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/endorsement-provisioning.yaml:19]: Need to incline the schema!

File: docs/api/endorsement-provisioning/endorsement-provisioning.yaml (hunk 24 above)
Comment (yogeshbdeshpande) [reply docs/api/endorsement-provisioning/endorsement-provisioning.yaml:5]: Add a description and fix the URL

**Comment by thomas-fossati** [comment]
Partially addresses #60

**Review by thomas-fossati — APPROVED** [review]
ship it!

**Pull request merged**